### PR TITLE
WIP: add AWS Cognito to Authproviders

### PR DESCRIPTION
**Pull request by josuablejeru**
# ⚠️ This is still Work in progress ⚠️ 

As suggested from @thedavidprice and @ajcwebdev  in issue #3452 I want to present to you all my first iteration on the AWS Auth provider.
Currently, you get an error in the console if you want to perform a login which I didn't now how to solve currently. Maybe someone has an Idea.
image

Also I wanted to add what the auth [setup auth configuration](https://github.com/redwoodjs/redwood/tree/main/packages/cli/src/commands/setup/auth/providers) was missing in the README (which is totally fine at the moment)

## Not included
- Documentation
- Role checks are not implemented yet

### CODE REVIEW

**Comment by dthyresson** [comment]
@josuablejeru Also, check you want to use `login: ({ username, password }) => {` that is `login` lowercase instead of `logIn` pascal case (capital `In`).

File: packages/auth/src/authClients/cognito.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+export const cognito = (client: CognitoUserPool): CognitoAuthClient => {
+  const poolData = {
+    UserPoolId: process.env?.COGNITO_USERPOOL_ID || '',
+    ClientId: process.env?.COGNITO_CLIENT_ID || '',
+  }
+  const userPool: CognitoUserPool = new CognitoUserPool(poolData)
+  return {
+    client: client,
+    type: 'cognito',
+    login: ({ username, password }) => {
```
[the remark this answers is not in the document]
Comment (josuablejeru) [reply packages/auth/src/authClients/cognito.ts:46]: @dthyresson I will take care to include them. 😄

**Comment by jtoar** [comment]
Hey @josuablejeru, I'm going to close this for now but let us know if you're still interested and we're happy to reopen and prioritize.

**Comment by joconor** [comment]
Any update on where this work stands? AWS Cognito is likely to become a priority for me within the next few months. Would be great if the core rwjs team would make this a priority, or if the core team could manage/shepherd community work toward providing AWS Cognito as an official auth provider

**Comment by jtoar** [comment]
Hey @joconor I don't think the core team will have time to prioritize it in the near future but we're more than happy to provide guidance and reviews.

**Comment by ajcwebdev** [comment]
Here are some resources that may be helpful.

### Official Cognito resources

* [Amazon Cognito Identity SDK for JavaScript](https://github.com/aws-amplify/amplify-js/tree/master/packages/amazon-cognito-identity-js)
* [Amazon Cognito user pools Auth API reference](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-userpools-server-contract-reference.html)
* [Amplify Authentication](https://docs.amplify.aws/cli/auth/overview/)
* [SSR Support for AWS Amplify JavaScript Libraries](https://aws.amazon.com/de/blogs/mobile/ssr-support-for-aws-amplify-javascript-libraries/)
* [Amplify UI `Authenticator` component](https://ui.docs.amplify.aws/components/authenticator)

It looks like there are potentially a few ways it could be integrated depending on whether you want to use Cognito by itself or if we want to try and integrate more heavily with the Amplify suite of tools.

### Examples from other frameworks

* [React <> Cognito](https://www.xtages.com/blog/posts/react-and-cognito/)
* [Gatsby + Cognito Authentication](https://www.bengladwell.com/gatsby-site-with-authentication/)
* [Amazon Cognito with NextAuth.js](https://next-auth.js.org/providers/cognito)
* [Next.js authentication with Next Auth and AWS Cognito](https://darrenwhite.dev/blog/nextjs-authentication-with-next-auth-and-aws-cognito)
* [Cognito authentication method for Remix](https://gist.github.com/chanan/d1601c63df36ac476e84fbc2ca96b2e9)

It looks like @ErikCH has been [doing some work in this area with Remix](https://github.com/remix-run/remix/issues/806) and I had proposed an integration a [long time ago](https://github.com/aws-amplify/docs/issues/2583) that could be worth revisiting.

**Comment by ErikCH** [comment]
Thanks @ajcwebdev !

Let me know if you have some issues with using Amplify with RedwoodJS. I haven't tried it, but I don't see why a user couldn't use it. We support React with our [JS library](https://docs.amplify.aws/start/q/integration/react/?sc_icampaign=react-start&sc_ichannel=docs-home)/

**Pull request closed without merging**